typed/union_test.go

```diff
@@ -0,0 +1,224 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package typed_test
+
+import (
+	"testing"
+
+	"sigs.k8s.io/structured-merge-diff/typed"
+)
+
+var unionParser = func() typed.ParseableType {
+	parser, err := typed.NewParser(`types:
+- name: union
+  struct:
+    fields:
+    - name: discriminator
+      type:
+        scalar: string
+    - name: a
+      type:
+        scalar: numeric
+    - name: b
+      type:
+        scalar: numeric
+    - name: c
+      type:
+        scalar: numeric
+    - name: one
+      type:
+        scalar: numeric
+    - name: two
+      type:
+        scalar: numeric
+    - name: three
+      type:
+        scalar: numeric
+    unions:
+      - discriminator: discriminator
+        fields: [a, b, c]
+      - fields: [one, two, three]`)
+	if err != nil {
+		panic(err)
+	}
+	return parser.Type("union")
+}()
+
+func TestNormalizeUnions(t *testing.T) {
+	tests := []struct {
+		name string
+		old  typed.YAMLObject
+		new  typed.YAMLObject
+		out  typed.YAMLObject
+	}{
+		{
+			name: "nothing changed, nothing changes",
+			old:  `{"a": 1}`,
+			new:  `{"a": 1}`,
+			out:  `{"a": 1}`,
+		},
+		{
+			name: "proper union update, setting discriminator",
+			old:  `{"a": 1}`,
+			new:  `{"b": 1}`,
+			out:  `{"b": 1, "discriminator": "b"}`,
+		},
+		{
+			name: "proper union update, no discriminator",
+			old:  `{"one": 1}`,
+			new:  `{"two": 1}`,
+			out:  `{"two": 1}`,
+		},
+		{
+			name: "proper union update from not-set, setting discriminator",
+			old:  `{}`,
+			new:  `{"b": 1}`,
+			out:  `{"b": 1, "discriminator": "b"}`,
+		},
+		{
+			name: "proper union update from not-set, no discriminator",
+			old:  `{}`,
+			new:  `{"two": 1}`,
+			out:  `{"two": 1}`,
+		},
+		{
+			name: "remove union, with discriminator",
+			old:  `{"a": 1}`,
+			new:  `{}`,
+			out:  `{}`,
+		},
+		{
+			name: "remove union and discriminator",
+			old:  `{"a": 1, "discriminator": "a"}`,
+			new:  `{}`,
+			out:  `{}`,
+		},
+		{
+			name: "remove union, not discriminator",
+			old:  `{"a": 1, "discriminator": "a"}`,
+			new:  `{"discriminator": "a"}`,
+			out:  `{"discriminator": "a"}`,
+		},
+		{
+			name: "remove union, no discriminator",
+			old:  `{"two": 1}`,
+			new:  `{}`,
+			out:  `{}`,
+		},
+		{
+			name: "dumb client update, no discriminator",
+			old:  `{"one": 1}`,
+			new:  `{"one": 2, "two": 1}`,
+			out:  `{"two": 1}`,
+		},
+		{
+			name: "dumb client update, sets discriminator",
+			old:  `{"a": 1}`,
+			new:  `{"a": 2, "b": 1}`,
+			out:  `{"b": 1, "discriminator": "b"}`,
+		},
+		{
+			name: "dumb client doesn't update discriminator",
+			old:  `{"a": 1, "discriminator": "a"}`,
+			new:  `{"a": 2, "b": 1, "discriminator": "a"}`,
+			out:  `{"b": 1, "discriminator": "b"}`,
+		},
+		{
+			name: "client sends multiple with discriminator",
+			old:  `{}`,
+			new:  `{"a": 1, "b": 1, "discriminator": "a"}`,
```

Comment:
I also think this should cause an error.

Reply:
Actually, the use-case in mind here is the following:
- I'm using kustomize
- my base deployment uses an emptyDir
- my overlay uses a persistent disk
- kustomize typically merges this by having both (you have to explicitely set `emptyDir` to `null`)

In that case, you could use the discriminator to clarify (which is much better than setting discriminator to null)

Reply:
It sounds like kustomize should use this library to do the merges. It'd be pretty nifty if customize produced output which had each overlay as a separate field manager.

Reply:
So this would mean someone could send 
```
deploymentStrategyType: Recreate
rollingUpdate:
  maxUnavailable: 1
```
and it would be accepted and treated as this now?
```
deploymentStrategyType: Recreate
rollingUpdate: {}
```